src/compiler/checker.ts

```diff
@@ -6638,6 +6686,412 @@ namespace ts {
             }
         }
 
+
+        function checkJsxSelfClosingElement(node: JsxSelfClosingElement) {
+            checkJsxOpeningLikeElement(node);
+            return jsxElementType;
+        }
+
+        function checkJsxElement(node: JsxElement) {
+            // Check that the closing tag matches
+            let expectedClosingTag = getTextOfNode(node.openingElement.tagName);
+            if (expectedClosingTag !== getTextOfNode(node.closingElement.tagName)) {
+                error(node.closingElement, Diagnostics.Expected_corresponding_JSX_closing_tag_for_0, expectedClosingTag);
+            }
+
+            // Check attributes
+            checkJsxOpeningLikeElement(node.openingElement);
+
+            // Check children
+            for (var i = 0, n = node.children.length; i < n; i++) {
+                if (node.children[i].kind === SyntaxKind.JsxExpression) {
+                    checkJsxExpression(<JsxExpression>node.children[i]);
+                }
+                else if (node.children[i].kind === SyntaxKind.JsxElement) {
+                    checkJsxElement(<JsxElement>node.children[i]);
+                }
+                else if (node.children[i].kind === SyntaxKind.JsxSelfClosingElement) {
+                    checkJsxSelfClosingElement(<JsxSelfClosingElement>node.children[i]);
+                }
+                else {
+                    // No checks for JSX Text
+                    Debug.assert(node.children[i].kind === SyntaxKind.JsxText);
+                }
+            }
+
+            return jsxElementType;
+        }
+
+        /// Returns true iff the JSX element name would be a valid identifier, ignoring restrictions about keywords not being identifiers
+        function isIdentifierLike(name: string) {
+            // - is the only character supported in JSX attribute names that isn't valid in JavaScript identifiers
+            return name.indexOf('-') < 0;
+        }
+
+        /// Returns true iff React would emit this tag name as a string rather than an identifier or qualified name
+        function isJsxIntrinsicIdentifier(tagName: Identifier|QualifiedName) {
+            if (tagName.kind === SyntaxKind.QualifiedName) {
+                return false;
+            }
+            else {
+                let firstChar = (<Identifier>tagName).text.charAt(0);
+                return firstChar.toLowerCase() === firstChar;
+            }
+        }
+
+        function checkJsxAttribute(node: JsxAttribute, elementAttributesType: Type, nameTable: Map<boolean>) {
+            var correspondingPropType: Type = undefined;

```

Comment:
Changed to `let`, but keeping the assignment to differentiate this from a "is initialized on all codepaths" variable